File: electron/bridge/__tests__/pairingQrV2.test.ts
/**
 * 阶段 D-02/D-04 服务层单测：
 * - BridgeService.getPairingInfo 增量 v2 字段（serverId/expiresAt/endpoints/capabilities），
 *   旧字段 host/port/fingerprint/expiresInSec 保持；
 * - getPairingQrPayload('v2' | 'legacy')：解析往返 + 旧格式兼容；
 * - 配对码一次性：消费后 v2 载荷中的 pairingCode 不能再被消费（扫码 -> 拒绝）；
 * - 过期二维码被拒绝（isPairingCodeValid / parsePairingQr expired）；
 * - mDNS TXT 字段（D-04）。
 */
import { describe, expect, it, vi, beforeEach, afterEach } from 'vitest'
import { rmSync } from 'node:fs'

vi.mock('electron', () => ({
  app: { getPath: () => '/tmp/qingyu-pairing-v2-test' },
  safeStorage: { isEncryptionAvailable: () => false },
  ipcMain: { handle: vi.fn(), on: vi.fn() },
  BrowserWindow: { getAllWindows: () => [] },
}))

import { BridgeService, getMachineFingerprint } from '../index'
import { buildMdnsTxt } from '../mdns'
import { getServerId, resetBridgeIdentityCache } from '../identity'
import { consumePairingCode, isPairingCodeValid } from '../auth'
import { parsePairingQr } from '../../../shared/pairingQr'

const TEST_ROOT = '/tmp/qingyu-pairing-v2-test'

beforeEach(() => {
  rmSync(TEST_ROOT, { recursive: true, force: true })
  resetBridgeIdentityCache()
})

afterEach(() => {
  rmSync(TEST_ROOT, { recursive: true, force: true })
})

describe('getPairingInfo v2 增量字段', () => {
  it('旧字段保留 + 新字段（serverId/displayName/apiVersion/capabilities/expiresAt/endpoints）', () => {
    const service = new BridgeService()
    const info = service.getPairingInfo()
    // 旧契约
    expect(typeof info.host).toBe('string')
    expect(info.port).toBeGreaterThan(0)
    expect(info.fingerprint).toMatch(/^[0-9a-f]{16}$/)
    expect(info.expiresInSec).toBeGreaterThan(0)
    expect(info.expiresInSec).toBeLessThanOrEqual(300)
    // v2 增量
    expect(info.serverId).toBe(getServerId())
    expect(typeof info.displayName).toBe('string')
    expect(info.apiVersion).toBe(1)
    expect(info.capabilities).toContain('pairing_qr_v2')
    expect(info.capabilities).toContain('settings_snapshot_v2')
    expect(typeof info.expiresAt).toBe('number')
    expect(info.expiresAt).toBeGreaterThan(Date.now())
    expect(info.endpoints?.length).toBeGreaterThanOrEqual(1)
    expect(info.endpoints?.[0]).toMatchObject({ host: info.host, port: info.port, security: 'LOCAL_CLEARTEXT' })
  })

  it('serverId 跨 BridgeService 实例（模拟重启）稳定；与 machineFingerprint 不同', () => {
    const a = new BridgeService().getPairingInfo()
    resetBridgeIdentityCache()
    const b = new BridgeService().getPairingInfo()
    expect(a.serverId).toBe(b.serverId)
    expect(a.serverId).not.toBe(getMachineFingerprint())
  })
})

describe('getPairingQrPayload（D-02）', () => {
  it('默认 v2：解析为 v2 且字段与 pairingInfo 一致', () => {
    const service = new BridgeService()
    const info = service.getPairingInfo()
    const parsed = parsePairingQr(service.getPairingQrPayload())
    expect(parsed.kind).toBe('v2')
    if (parsed.kind !== 'v2') return
    expect(parsed.payload.serverId).toBe(info.serverId)
    expect(parsed.payload.pairingCode).toBe(info.fingerprint)
    expect(parsed.payload.expiresAt).toBe(info.expiresAt)
    expect(parsed.payload.certificatePin).toBeNull()
    expect(parsed.payload.endpoints[0].host).toBe(info.host)
  })

  it('legacy 模式：旧 Android 可解析的 {host,port,fingerprint}', () => {
    const service = new BridgeService()
    const info = service.getPairingInfo()
    const raw = service.getPairingQrPayload('legacy')
    expect(JSON.parse(raw)).toEqual({ host: info.host, port: info.port, fingerprint: info.fingerprint })
    const parsed = parsePairingQr(raw)
    expect(parsed.kind).toBe('legacy')
  })

  it('配对码一次性：扫码消费后同一 QR 再次配对被拒（v2 pairingCode 失效）', () => {
    const service = new BridgeService()
    const parsed = parsePairingQr(service.getPairingQrPayload())
    expect(parsed.kind).toBe('v2')
    if (parsed.kind !== 'v2') return
    const code = parsed.payload.pairingCode
    expect(isPairingCodeValid(code)).toBe(true)
    // 首次配对消费成功
    expect(consumePairingCode(code)).toBe(true)
    // 重复使用同一二维码 -> 配对码已失效
    expect(consumePairingCode(code)).toBe(false)
    expect(isPairingCodeValid(code)).toBe(false)
  })

  it('自动轮换：消费后重新取码，旧 QR 过期内容拒绝、新 QR 有效', () => {
    const service = new BridgeService()
    const first = parsePairingQr(service.getPairingQrPayload())
    expect(first.kind).toBe('v2')
    if (first.kind !== 'v2') return
    consumePairingCode(first.payload.pairingCode)

    const second = parsePairingQr(service.getPairingQrPayload())
    expect(second.kind).toBe('v2')
    if (second.kind !== 'v2') return
    expect(second.payload.pairingCode).not.toBe(first.payload.pairingCode)
    expect(isPairingCodeValid(second.payload.pairingCode)).toBe(true)
  })

  it('伪造过期 v2 二维码：PC 侧解析即拒绝（expired）', () => {
    const service = new BridgeService()
    const parsed = parsePairingQr(service.getPairingQrPayload())
    expect(parsed.kind).toBe('v2')
    if (parsed.kind !== 'v2') return
    const stale = JSON.stringify({ ...parsed.payload, expiresAt: Date.now() - 1000 })
    expect(parsePairingQr(stale).kind).toBe('invalid')
  })
})

describe('mDNS TXT（D-04）', () => {
  it('buildMdnsTxt 字段齐全且为字符串（bonjour TXT 兼容）', () => {
    const txt = buildMdnsTxt('server-uuid', 1, 2, false, '我的电脑')
    expect(txt).toEqual({
      serverId: 'server-uuid',
      apiVersion: '1',
      pairVersion: '2',
      tls: '0',
      displayName: '我的电脑',
    })
  })
})

File: android/app/src/test/java/com/qingyu/companion/model/PairingQrV2Test.kt
package com.qingyu.companion.model

import com.qingyu.companion.network.NetworkModule
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

/** QR v2 解析测试（方案 §8 D-02/D-03：新旧格式兼容 + 逐项校验） */
class PairingQrV2Test {

    private val json = NetworkModule.json
    private val now = 1_788_000_000_000L

    private fun v2Json(
        version: String = "2",
        scheme: String = "\"qingyu-pair\"",
        expiresAt: String = (now + 300_000).toString(),
        endpoints: String = """[{"host":"192.168.1.8","port":8321,"security":"LOCAL_CLEARTEXT"}]""",
        pairingCode: String = "\"one-time-code\"",
    ) = """
        {
          "version": $version,
          "scheme": $scheme,
          "serverId": "uuid-1",
          "displayName": "我的电脑",
          "apiVersion": 1,
          "capabilities": ["settings_snapshot_v2", "task_events_v2"],
          "pairingCode": $pairingCode,
          "expiresAt": $expiresAt,
          "endpoints": $endpoints,
          "certificatePin": null
        }
    """.trimIndent()

    @Test
    fun `v2 完整载荷解析成功`() {
        val result = parsePairingQr(v2Json(), json, now)
        assertTrue(result is PairingQrResult.V2)
        val p = (result as PairingQrResult.V2).payload
        assertEquals("uuid-1", p.serverId)
        assertEquals("我的电脑", p.displayName)
        assertTrue("settings_snapshot_v2" in p.capabilities)
        assertEquals("192.168.1.8", p.firstUsableEndpoint()?.host)
        assertEquals(8321, p.firstUsableEndpoint()?.port)
        assertEquals("one-time-code", p.pairingCode)
    }

    @Test
    fun `旧格式 host-port-fingerprint 按 legacy 解析`() {
        val raw = """{"host":"10.0.0.5","port":8321,"fingerprint":"aa:bb"}"""
        val result = parsePairingQr(raw, json, now)
        assertTrue(result is PairingQrResult.Legacy)
        assertEquals("10.0.0.5", (result as PairingQrResult.Legacy).payload.host)
    }

    @Test
    fun `非 JSON 输入 Invalid`() {
        val result = parsePairingQr("not-json", json, now)
        assertTrue(result is PairingQrResult.Invalid)
        assertEquals(PairingQrInvalidReason.NOT_JSON, (result as PairingQrResult.Invalid).reason)
    }

    @Test
    fun `不支持的 version 拒绝而非误判 legacy`() {
        val result = parsePairingQr(v2Json(version = "3"), json, now)
        assertEquals(PairingQrInvalidReason.VERSION_UNSUPPORTED, (result as PairingQrResult.Invalid).reason)
    }

    @Test
    fun `scheme 不匹配拒绝`() {
        val result = parsePairingQr(v2Json(scheme = "\"other\""), json, now)
        assertEquals(PairingQrInvalidReason.SCHEME_MISMATCH, (result as PairingQrResult.Invalid).reason)
    }

    @Test
    fun `过期二维码拒绝`() {
        val result = parsePairingQr(v2Json(expiresAt = (now - 1).toString()), json, now)
        assertEquals(PairingQrInvalidReason.EXPIRED, (result as PairingQrResult.Invalid).reason)
    }

    @Test
    fun `expiresAt 为 0 视为不过期`() {
        val result = parsePairingQr(v2Json(expiresAt = "0"), json, now)
        assertTrue(result is PairingQrResult.V2)
    }

    @Test
    fun `无有效 endpoint 拒绝（host 空或端口越界）`() {
        val badHost = v2Json(endpoints = """[{"host":"","port":8321}]""")
        assertEquals(
            PairingQrInvalidReason.NO_VALID_ENDPOINT,
            (parsePairingQr(badHost, json, now) as PairingQrResult.Invalid).reason,
        )
        val badPort = v2Json(endpoints = """[{"host":"1.2.3.4","port":0}]""")
        assertEquals(
            PairingQrInvalidReason.NO_VALID_ENDPOINT,
            (parsePairingQr(badPort, json, now) as PairingQrResult.Invalid).reason,
        )
    }

    @Test
    fun `pairingCode 缺失拒绝`() {
        val result = parsePairingQr(v2Json(pairingCode = "\"\""), json, now)
        assertEquals(PairingQrInvalidReason.MISSING_PAIRING_CODE, (result as PairingQrResult.Invalid).reason)
    }

    @Test
    fun `多 endpoint 取第一个可用并跳过脏项`() {
        val raw = v2Json(
            endpoints = """[{"host":"","port":1},{"host":"192.168.1.9","port":8322,"security":"TLS_SYSTEM"}]""",
        )
        val p = (parsePairingQr(raw, json, now) as PairingQrResult.V2).payload
        assertEquals("192.168.1.9", p.firstUsableEndpoint()?.host)
        assertEquals(8322, p.firstUsableEndpoint()?.port)
    }

    @Test
    fun `旧格式字段不全归为 LEGACY_INCOMPLETE`() {
        val result = parsePairingQr("""{"host":"1.2.3.4"}""", json, now)
        assertTrue(result is PairingQrResult.Invalid)
        assertEquals(PairingQrInvalidReason.LEGACY_INCOMPLETE, (result as PairingQrResult.Invalid).reason)
    }
}
